src/masternodeman.cpp

```diff
@@ -644,6 +671,401 @@ void CMasternodeMan::ProcessMessage(CNode* pfrom, std::string& strCommand, CData
+                        pair.first, (int)vecMasternodeRanks.size(), MAX_POSE_RANK);
+            break;
+        }
+    }
+    LogPrint("masternode", "CMasternodeMan::DoFullVerificationStep -- Sent verification requests to %d masternodes\n", nCount);
+}
+
+void CMasternodeMan::CheckSameAddr()
+{
+    if(!masternodeSync.IsSynced() || vMasternodes.empty()) return;
+
+    std::vector<CMasternode*> vBan;
+    std::vector<CMasternode*> vSortedByAddr;
+    std::set<CAddress> setAddrsToVerify;
+
+    size_t nMaxAddrToVerify = 10;
+
+    {
+        LOCK(cs);
+
+        CMasternode* prevMasternode;
+
+        BOOST_FOREACH(CMasternode& mn, vMasternodes) {
+            vSortedByAddr.push_back(&mn);
+        }
+
+        sort(vSortedByAddr.begin(), vSortedByAddr.end(), CompareByAddr());
+
+        BOOST_FOREACH(CMasternode* pmn, vSortedByAddr) {
+            // check only (pre)enabled masternodes
+            if(!pmn->IsEnabled() && !pmn->IsPreEnabled()) continue;
+            // initial step
+            if(!prevMasternode) {
+                prevMasternode = pmn;
+                continue;
+            }
+            // second+ step
+            if(pmn->addr == prevMasternode->addr) {
+                if(prevMasternode->IsPoSeVerified()) {
+                    vBan.push_back(pmn);
+                } else if(pmn->IsPoSeVerified()) {
+                    vBan.push_back(prevMasternode);
+                    prevMasternode = pmn;
+                } else {
+                    // neither one of them is verified, let's try now
+                    setAddrsToVerify.insert((CAddress)pmn->addr);
+                    if(setAddrsToVerify.size() >= nMaxAddrToVerify) break;
+                }
+            }
+        }
+    }
+
+    // ban duplicates
+    BOOST_FOREACH(CMasternode* pmn, vBan) {
+        pmn->nPoSeBanScore++;
+    }
+
+    BOOST_FOREACH(const CAddress &addr, setAddrsToVerify) {
+        LogPrintf("CMasternodeMan::CheckSameAddr -- Multiple Masternodes with the same IP found addr=%s\n", addr.ToString());
+        SendVerifyRequest(addr, vSortedByAddr);
+    }
+}
+
+bool CMasternodeMan::SendVerifyRequest(const CAddress& addr, const std::vector<CMasternode*>& vSortedByAddr)
+{
+    CNode* pnode = ConnectNode(addr, NULL, true);
+    if(pnode != NULL) {
+        if(netfulfilledman.HasFulfilledRequest(addr, strprintf("%s", NetMsgType::MNVERIFY)+"-request")) {
+            // we already asked for verification, not a good idea to do this too often, skip it
+            LogPrint("masternode", "CMasternodeMan::SendVerifyRequest -- too many requests, skipping... addr=%s\n", addr.ToString());
+            return false;
+        } else {
+            netfulfilledman.AddFulfilledRequest(addr, strprintf("%s", NetMsgType::MNVERIFY)+"-request");
+            // use random nonce, store it and require node to reply with correct one later
+            CMasternodeVerification mnv(addr, 1 + insecure_rand() % 999999, pCurrentBlockIndex->nHeight);
```

Comment:
Is it possible that the target node will get verify request msg before the latest block be accepted ?